src/router/index.ts

```diff
@@ -0,0 +1,422 @@
+
+// This is only respected by Safari.
+const documentTitle = "Radicle Interface";
+
+export const historyStore = writable<Route[]>([BOOT_ROUTE]);
+export const activeRouteStore: Writable<Route> = writable(BOOT_ROUTE);
+
+export function link(node: any, route?: Route) {
+  function onClick(event: any): void {
+    const anchor = event.currentTarget;
+    let path: string | Route;
+
+    // Checking if we passed a route param into use:link
+    if (!route) {
+      const url = new URL(anchor.href);
+      path = url.pathname + url.search + url.hash;
+    } else {
+      path = route;
+    }
+
+    if (anchor.target === "") {
+      event.preventDefault();
+      navigate(path, {
+        replace: anchor.hasAttribute("replace"),
+      });
+    }
+  }
+
+  node.addEventListener("click", onClick);
+
+  return {
+    destroy() {
+      node.removeEventListener("click", onClick);
+    },
+  };
+}
+
+// Replaces history on any user interaction with forward and backwards buttons
+// with the current window.history.state
+window.addEventListener("popstate", e => {
+  if (e.state) replace(e.state);
+});
+
+export function navigate(
+  route: Route | string,
+  opts?: { replace?: boolean },
+): void {
+  if (typeof route === "string") {
+    route = pathToRoute(route);
+  }
+
+  // If we don't explicitly pass a anchor tag hash or query string, we should reset it globally
+  if (route.type === "projects" && !route.params.hash) {
+    route.params = { ...route.params, hash: null, search: null };
+  }
+
+  // If activeRoute and newRoute are type === "projects" combine the route params
+  const activeRoute = get(activeRouteStore);
+  if (route.type === "projects" && activeRoute.type === "projects") {
+    route.params = { ...activeRoute.params, ...route.params };
+  }
+
+  if (opts?.replace) {
+    replace(route);
+  } else {
+    push(route);
+  }
+}
+
+export const push = (newRoute: Route): void => {
+  const history = get(historyStore);
+
+  // Limit history to a maximum of 10 steps. We shouldn't be doing more than
+  // one subsequent pop() anyway.
+  historyStore.set([...history, newRoute].slice(-10));
+  activeRouteStore.set(newRoute);
+  window.history.pushState(newRoute, documentTitle, routeToPath(newRoute));
+};
+
+export const pop = (): void => {
+  const history = get(historyStore);
+  const newRoute = history.pop();
+  if (newRoute) {
+    historyStore.set(history);
+    activeRouteStore.set(newRoute);
+    window.history.back();
+  }
+};
+
+function replace(newRoute: Route): void {
+  historyStore.set([newRoute]);
+  activeRouteStore.set(newRoute);
+  window.history.replaceState(newRoute, documentTitle, routeToPath(newRoute));
+}
+
+export const initialize = () => {
+  const { pathname, search, hash } = window.location;
+  replace(pathToRoute(pathname + search + hash));
+};
+
+export function pathToRoute(path: string | null): Route {
+  if (!path) {
+    return { type: "404" };
+  }
+
+  const url = new URL(path, "https://app.radicle.xyz");
```

Comment:
Does this work in local development?

Reply:
Yeah this works in a local dev environment, since we don't use the base url, but eventually we could use `window.origin` here. 🤔

Reply:
fixed in 9fe92d1